src/sparse/vec.rs

```diff
@@ -961,11 +976,84 @@ where IS: Deref<Target=[usize]>,
     }
 }
 
+impl<N: Num + Copy, I: SpIndex> Zero for CsVecI<N, I> {
+    fn zero() -> CsVecI<N, I> {
+        CsVecI::new(0, vec![], vec![])
+    }
+
+    fn is_zero(&self) -> bool {
+        self.data.iter().all(|x| x.is_zero())
+    }
+}
+
+#[cfg(feature = "alga")]
+mod alga_impls {
+    use super::*;
+    use alga::general::*;
+
+    impl<N: Clone + Copy + Num, I: Clone + SpIndex> AbstractMagma<Additive> for CsVecI<N, I> {
+        fn operate(&self, right: &CsVecI<N, I>) -> CsVecI<N, I> {
+            self + right
+        }
+    }
+
+    impl<N: Copy + Num, I: SpIndex> Identity<Additive> for CsVecI<N, I> {
+        fn identity() -> CsVecI<N, I> {
+            CsVecI::zero()
+        }
+    }
+
+    impl<N: Copy + Num, I: SpIndex> AbstractSemigroup<Additive> for CsVecI<N, I> {}
+
+    impl<N: Copy + Num, I: SpIndex> AbstractMonoid<Additive> for CsVecI<N, I> {}
+
+    impl<N, I> Inverse<Additive> for CsVecI<N, I>
+        where N: Clone + Neg<Output=N> + Copy + Num,
+              I: SpIndex
+    {
+        fn inverse(&self) -> CsVecI<N, I> {
+            -self.clone()
```

Comment:
Am I right in assuming this will create a zero element and then subtract `self`from it? I'm wondering if it wouldn't be better to explicitly return a new vector like this:

```rust
CsVec {
    data: self.data.iter().map(|x| -x).collect(),
    indices: self.indices.clone(),
    dim: self.dim
}
```

But it's possible that your version compiles down to the same code.

Reply:
No, this delegates to the `Neg` implementation, which modifies the values in-place, as it takes the vector by value. So all `inverse` does is to clone the vector and to negate its `data` attribute element-wise in-place. There is a copying vs. cache locality trade-off, but we are not going to save an allocation that way.

So… I did some benchmarks, your suggestion is indeed about 20% faster for a vector with about 10k entries. I am adding the change in a commit including the benchmarks.